tf2/src/cache.cpp

```diff
@@ -255,14 +255,26 @@ bool TimeCache::insertData(const TransformStorage & new_data)
   }
 
   // Find the oldest element in the list before the incoming stamp.
-  auto last_transform_pos = std::find_if(
+  auto insertion_pos = std::find_if(
     storage_.begin(), storage_.end(), [&](const auto & transfrom) {
       return transfrom.stamp_ <= new_data.stamp_;
     });
 
+  bool should_insert = true;
+  // Search along all data with same timestamp (sorted), and only insert if we
+  // did not find the exact same data.
+  auto maybe_same_pos = insertion_pos;
+  while (maybe_same_pos != storage_.end() && maybe_same_pos->stamp_ == new_data.stamp_) {
```

Comment:
If the benchmark says it's better, I wouldn't ever worry. But as a matter of curiosity, giving a quick look to the `std::find` implementation from GCC, how is this any different?
```
  template<typename _InputIterator, typename _Predicate>
    _GLIBCXX20_CONSTEXPR
    inline _InputIterator
    __find_if(_InputIterator __first, _InputIterator __last,
	      _Predicate __pred, input_iterator_tag)
    {
      while (__first != __last && !__pred(__first))
	      ++__first;
      return __first;
    }
```

Reply:
Right, I agree with making improvements here.

For what it is worth, we are still iterating over parts (and in the worst case, all) of the list 3 times here:

1.  When we find the insertion point (which may be at the end).
2.  When we actually insert.
3.  When we prune the list.

I think it would be nice to find a way to reduce this to only one iteration over the loop, though I'm not 100% sure that is possible.  I wouldn't call this a blocker for getting this in, but since we are trying to improve the performance here I think it would make sense to go as far as we can.

@EricCousineau-TRI would you mind taking a look and seeing if we can do that?

Reply:
Off the top of my head, I don't know how to easily blend the loops beyond a more intense refactoring, that I'm not 100% sure on.

That being said, we can somewhat reduce how much iteration is done by pruning first, so we don't iterate over unused data. I'll add that in.
There may be a behavior change, so we'll see if it affects downstream behavior.

Reply:
Updated `pruneList()` 215d047e

Reply:
Reasoning out loud:
- We now prune all old data (and should never "reconsider" it)
- We use `find_if` to scan from newest to oldest; we stop at the first point where timestamp is equal (or older)
- From this point, we scan with more expensive full equality check to ensure we have no duplicates.

I believe this should only ever scan at most N times, never more.

Does this reason seem sound?

Reply:
Yep, your reasoning makes sense to me!

I'm going to suggest that we add a comment near the top of this function explaining exactly this, so that future readers (us!) will know that we thought about this.

Reply:
Done 6571fe9e

Reply:
This looks great, thanks.

I'm going to run CI on it one more time.